syntax = "proto3";

option go_package = "github.com/zhangjyr/distributed-notebook/common/gateway";
option java_multiple_files = true;
option java_package = "com.github.zhangjyr.distributed-notebook.common.gateway";
option java_outer_classname = "GatewayProto";

package gateway;

// The juypter gateway service for host local kernels.
service LocalGateway {
  // Start a kernel or kernel replica.
  rpc StartKernel(KernelSpec) returns (KernelConnectionInfo) {}

  // GetKernelStatus returns the status of a kernel.
  rpc GetKernelStatus(KernelId) returns (KernelStatus) {}

  // KillKernel kills a kernel.
  rpc KillKernel(KernelId) returns (Void) {}

  // StopKernel stops a kernel gracefully and return immediately.
  rpc StopKernel(KernelId) returns (Void) {}

  // WaitKernel waits for a kernel to stop and return status.
  rpc WaitKernel(KernelId) returns (KernelStatus) {}
}

// The input for starting a kernel replica.
message LocalKernelSpec {
  string replicaId = 1;
  repeated string replicas = 2;
  KernelSpec spec = 3;
}

// The kernel id.
message KernelId {
  string id = 1;
}

// The kernel spec.
message KernelSpec {
  string id = 1;  // The kernel id.
  repeated string argv = 2;
	string signatureScheme = 3; // The signature scheme to use.
  string key = 4; // The key to use for signing messages.
}

// The connection info for a kernel.
message KernelConnectionInfo {
	string ip = 1; // The IP address of the kernel.
  string transport = 2; // The transport protocol to use.
  int32 controlPort = 3; // The port for control messages.
	int32 shellPort = 4; // The port for shell messages.
	int32 stdinPort = 5; // The port for stdin messages.
	int32 hbPort = 6; // The port for heartbeat messages.
	int32 iopubPort = 7; // The port for iopub messages.
	string signatureScheme = 8; // The signature scheme to use.
	string key = 9;  // The key to use for signing messages.
}

// The status of a kernel.
message KernelStatus {
  int32 status = 1;
}

message Void{}